### ISSUE TITLE
Unable to build sample (shader program linkage fails)

### ISSUE THREAD

**Issue by jaredpetker** [servo/webrender#594]
While trying to build the webrender sample, running this command

```
cargo run ../webrender/res
```

I get these warnings / errors

```
Failed to link shader program: WARNING: Could not find vertex shader attribute 'aColorTexCoordRectBottom' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aColorRectTL' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aMaskTexCoordRectTop' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aColorTexCoordRectTop' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aColorRectBL' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aColorRectTR' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aMaskTexCoordRectBottom' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aBorderPosition' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aBlurRadius' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'aBord
thread 'main' panicked at '-- Program link failed - exiting --', /Users/jaredpetker/dev/me/applesauce/hacksaw/webrender/webrender/src/device.rs:415
stack backtrace:
   1:        0x10134e2ea - std::sys::imp::backtrace::tracing::imp::write::h944c02ac40aee2d7
   2:        0x10135039f - std::panicking::default_hook::{{closure}}::h6875a2976258b020
   3:        0x101350047 - std::panicking::default_hook::h88ffbc5922643264
   4:        0x101350866 - std::panicking::rust_panic_with_hook::ha5aed1dfc0e220e3
   5:        0x100fd3253 - std::panicking::begin_panic::h5f0a1fd8f4de57ec
   6:        0x1011ab0f1 - webrender::device::Program::attach_and_bind_shaders::h2f9efeddfcb039d9
   7:        0x1011aff06 - webrender::device::Device::load_program::h8c077dd24fccb683
   8:        0x1011af4e8 - webrender::device::Device::create_program_with_prefix::h604f88449d78b22c
   9:        0x1011f9849 - webrender::renderer::create_prim_shader::h81fca8a8aa88c6f0
  10:        0x1011f8ab9 - webrender::renderer::LazilyCompiledShader::get::h8431e0cd91bde5af
  11:        0x1011f8899 - webrender::renderer::LazilyCompiledShader::new::hb86ab17d51b82a7c
  12:        0x1011f91b2 - webrender::renderer::PrimitiveShader::new::hfea1fc8232578c14
  13:        0x1011fb285 - webrender::renderer::Renderer::new::he761225cf7cdf8e2
  14:        0x100ea7603 - wr_sample::main::hdb4e48c7eb366261
  15:        0x1013516ba - __rust_maybe_catch_panic
  16:        0x101350c06 - std::rt::lang_start::h5d71a3afaaa4b2ff
  17:        0x100ead2d9 - main
```

I feel I must be doing something wrong since this project is actively worked on, however it seems like some of the vertex attributes (their accompanying shaders) were deleted a few weeks ago. What am I missing?

**Comment by kvark** [comment]
Since #563, the shader sources are included in the executables. So all you need to do is `cargo run` for the sample/replay tools. Passing the parameter of the shader path should work as an override now, but there was a bug with the error like you get, later fixed in #570. Perhaps, you have the source checked out between those PRs so that you have a bug? Try getting the latest. I've checked locally - `cargo run ../webrender/res` works as expected. cc @vvuk

**Comment by jaredpetker** [comment]
I did notice I didn't need to use the directory with `cargo run` post posting things after going through logs looking for glsl change, however I get the same error either way, I just pulled last night, and am on HEAD as far as I can tell as of this posting:

```
commit 655cbfc5a5de1d7bdce52205449779f912b4df6d
Merge: c625672 c33d8e4
Author: bors-servo <lbergstrom+bors@mozilla.com>
Date:   Fri Nov 25 01:16:37 2016 -0800
```

I feel like the warnings / error are warranted, as I can't find those gl attributes within the shader files to even be bound to.  Do you have shader files with the above attributes in them (such as `aColorTexCoordRectBottom `)?  I've searched and don't have that attribute defined anywhere in the shader files in what I've pulled, (but again, perhaps I'm missing something).

**Comment by kvark** [comment]
I find it weird that it failed to link the shaders just because there are warnings. What's your Mesa version, OS, and the HW/driver?
I'm also not seeing where stuff like `aColorTexCoordRectBottom` is used. Perhaps, you can clean up those unused attributes, make it compile, and file a PR?

**Comment by jaredpetker** [comment]
osmesa-sys: 0.1.2
osx: 10.11.6 el capitan

I'm starting to also believe the issues are perhaps unrelated to the warnings, poking around a bit here.

**Comment by jaredpetker** [comment]
So I notice the sample tries to pick a specific opengl version (3.2), but doesn't seem to actually get used, instead, `OpenGL version 4.1 INTEL-10.14.73` is used.

**Comment by kvark** [comment]
That by itself is fine. The context received has to be compatible with the one we ask, but doesn't have to have exactly the same GL version.

**Comment by jaredpetker** [comment]
I think my log was actually getting cut off, after commenting out the things causing warnings, I've gotten this:

```
ERROR: Input of fragment shader 'vClipMaskUvBounds' not written by vertex shader
ERROR: Input of fragment shader 'vClipMaskUv' not written by vertex shader
```

caused by loading 'prim_shared.glsl', going down this path

**Comment by kvark** [comment]
Interesting. What shader is that exactly?

**Comment by jaredpetker** [comment]
Ok so I've cornered the issue a bit, when a new Renderer is created it starts trying to link a bunch of shaders, once it starts doing a PrimitiveShader, (first one being `ps_rectangle`) I would see this error.

If I go into prim_shader.glsl and comment out the insides of the `write_clip` and `do_clip` functions, then the sample runs at least (of course, something is still awry). Basically I just did this to test.

write_clip
```
void write_clip(vec2 global_pos, ClipArea area) {
//    vec2 texture_size = textureSize(sCache, 0).xy;
//    vec2 uv = global_pos + area.task_bounds.xy - area.screen_origin_target_index.xy;
//    vClipMaskUvBounds = area.task_bounds / texture_size.xyxy;
//    vClipMaskUv = vec3(uv / texture_size, area.screen_origin_target_index.z);
}
```

do_clip
```
float do_clip() {
    // anything outside of the mask is considered transparent
//    bvec4 inside = lessThanEqual(
//        vec4(vClipMaskUvBounds.xy, vClipMaskUv.xy),
//        vec4(vClipMaskUv.xy, vClipMaskUvBounds.zw));
//    // check for the dummy bounds, which are given to the opaque objects
//    return vClipMaskUvBounds.xy == vClipMaskUvBounds.zw ? 1.0:
//        all(inside) ? textureLod(sCache, vClipMaskUv, 0).a : 0.0;
    return 1.0;
}
```

While at least I have things running for what I was interested in messing with, can probably try and figure out the issue here (unless someone knows off the top of their head and I can verify).

**Comment by kvark** [comment]
Hmm, for `ps_rectangle` the clip functions are supposed to be gated by `WR_FEATURE_CLIP`.

**Comment by jaredpetker** [comment]
Well they aren't for me, can add and put a pr in if that's what it was meant to be.

**Comment by jaredpetker** [comment]
Oh sorry, they are, missed what you actually meant ;)

**Comment by kvark** [comment]
Please try wrapping the `do_clip()` and `write_clip` functions into `#ifdef WR_FEATURE_CLIP`

**Comment by jaredpetker** [comment]
I went back to confirm that it was `ps_rectangle`, and not the subsequent `ps_rectangle_with_clipping`, and it is indeed `ps_rectangle_with_clipping` where the error occurs, so the guards are working (sorry for the mixup).

**Comment by jaredpetker** [comment]
Back with a fresh pair of eyes, `ps_yuv_image.fs.glsl` seems to be a culprit, it tries to do `do_clip`, but never does a `write_clip` in the vertex shader.

**Comment by kvark** [comment]
Please file a PR if you can - should be an easy fix.

**Comment by jaredpetker** [comment]
Will give it a go 👍

**Comment by kvark** [comment]
Fixed by #598
